rustls/src/msgs/message.rs

```diff
@@ -144,13 +124,75 @@ impl Message {
             .ok_or(MessageError::TooShortForLength)?;
         let payload = Payload::read(&mut sub);
 
-        Ok(Message {
+        Ok(RawMessage {
             typ,
             version,
-            payload: MessagePayload::Opaque(payload),
+            payload,
         })
     }
 
+    pub fn decode_payload(self) -> Result<Message, (Message, Error)> {
+        let (payload, err) = match MessagePayload::new(self.typ, self.version, self.payload) {
+            Ok(payload) => (payload, None),
+            Err((payload, err)) => (payload, Some(err)),
+        };
+
+        let msg = Message {
+            typ: self.typ,
+            version: self.version,
+            payload,
+        };
+
+        match err {
+            None => Ok(msg),
+            Some(e) => Err((msg, e)),
+        }
+    }
+
+    pub fn borrow(&self) -> BorrowMessage<'_> {
+        BorrowMessage {
+            typ: self.typ,
+            version: self.version,
+            payload: &self.payload.0,
+        }
+    }
+
+    pub fn is_content_type(&self, typ: ContentType) -> bool {
+        self.typ == typ
+    }
+}
+
+/// A TLS frame, named TLSPlaintext in the standard.
+/// This type owns all memory for its interior parts.
```

Comment:
Again, the documentation should be clarified so it is clear what the difference is between a `RawMessage` and a `Message`.

Reply:
Clarified.